src/pages/node-express/lesson13-a1.mdx

```diff
@@ -0,0 +1,526 @@
+          } else {
+            return done(null, false, { message: "Incorrect credentials." });
+          }
+        } catch (e) {
+          return done(e);
+        }
+      }
+    )
+  );
+
+  passport.serializeUser(async function (user, done) {
+    done(null, user.id);
+  });
+
+  passport.deserializeUser(async function (id, done) {
+    try {
+      const user = await User.findById(id);
+      if (!user) {
+        return done(new Error("user not found"));
+      }
+      return done(null, user);
+    } catch (e) {
+      done(e);
+    }
+  });
+};
+
+module.exports = passportInit;
+```
+
+Here we are registering a "strategy" to tell Passport how to handle requests that we pass to it.
+`passport.use` is a function that takes the name you want to give this passport strategy (here we're using `"local"`) and
+the strategy as the second argument, where we're passing in a `new LocalStrategy` from the `passport-local` library.
+
+The `LocalStrategy` constructor takes two arguments: an options object, and a callback function. The options object expects
+us to tell the strategy what field on the request will have the user identifier, and what field will have the password. Here, we tell it to look for `"email"` as the identifier, and `password` for the password.
+
+The second argument passed to the `LocalStrategy` constructor is the callback function that will be called when we do `passport.authenticate(...)` in the login route in `sessionRoutes.js`. The `local-strategy` library will extract the `email` and `password` fields that we defined
+in the options/first argument, from the `req.body`. If it doesn't find those fields, it will return a 400 error. If it does find them, then it will then pass those (along with a `done` callback) to the function we've defined above. It's up to us then to write the code specific to our application, that handles checking if the user exists and has used to correct password.
+
+The `done` callback accepts three arguments:
+
+1. An error, if there is one, otherwise just pass in `null`
+2. The user object, if the user is found and used valid credentials, otherwise `false`
+3. An object with a `message` property and `type`, this can be used in both error and success cases to show flash messages. Here we'e only using it in non-error failure cases (user not found or wrong credentials used).
+
+In addition to registering our local strategy, we also define `serializeUser` and `deserializeUser` functions on the `passport` object. `serializeUser` is used when saving a user to the request session object. We can't save an object to the session cookie, so we need to "serialize" it a.k.a. encode it as a string. In the above function, we're doing that by telling Passport to save just the user id to the session cookie. Then when Passport is handling a protected route, it will use the `deserializeUser` function to retrieve the user from the database, using the id that was saved to the session cookie.
```

Comment:
All new text to walk through how the Passport code in the last example works / when it actually gets called